src/OrleansRuntime/GrainTypeManager/GrainTypeData.cs

```diff
@@ -151,5 +154,11 @@ internal static PlacementStrategy GetPlacementStrategy(Type grainClass, Placemen
 
             return predicate;
         }
+
+        private ParameterInfo[] GetConstructorParameters()
+        {
+            var constructor = this.Type.GetConstructors(BindingFlags.Public | BindingFlags.Instance).FirstOrDefault();
+            return constructor?.GetParameters() ?? EmptyParameters;
+        }
```

Comment:
> The facet system design we came up with (and agreed on) was that features would be injected in constructors with metadata provided by attributes. How is this to be done without constructor selection?

I need to be more clear: I like this feature *very much*, and I like the pattern with ctor parameter attributes for wiring it up. The question about `IGrainActivator` does need to be answered, though, because this system does place additional constraints on an otherwise very open-ended API. I don't think that we should allow this facets pattern to be optionally supported, if we are going down this road then it should be mandatory that `IGrainActivator` follows consistent ctor selection and doesn't allow custom registrations. It's enough that we tell implementers in doc comments that they *must* adhere to this and that the easiest way to do so is to construct their types using `ActivatorUtils.CreateFactory`. It should be mandatory, since the intention is that this is a central extensibility pattern.

> Regarding the GrainTypeData. It is serializable so grain type data can be sent to the client and other silos (heterogeneous deployments), but the constructor arguments are only ever required locally, so there is no reason to serialize them.

I won't push back, but note that we have had problems with this in the past.